### DISCUSSION TITLE
S3Key of bundled lambda asset changes on every run in Github Actions workflow

### DISCUSSION THREAD

**Discussion by fynnfluegge** [aws/aws-cdk#25410] [Q&A]
Hi 👋,

I recently moved my cdk deployment to a workflow in github actions.

But on every run, all my lambda functions are updated during `cdk deploy` (even there is no change), since the S3key and the asset path of the new bundled asset differs on every run with the deployed version:

![Screenshot 2023-05-03 at 13 34 28](https://user-images.githubusercontent.com/16321871/235904913-fdbe59fd-d706-4451-9ef5-d3c50aba7f97.png)

I expect that only the lambda functions, where the code has changed will be updated. I bet this has something to do with the github actions runner that causes some minor tweaks on every run which leads to a different hash of the asset path and as a result a different S3key.

Has someone bypassed this issue?

Thanks in advance!

**Comment by peterwoodworth** [comment]
> I bet this has something to do with the github actions runner that causes some minor tweaks on every run which leads to a different hash of the asset path and as a result a different S3key.

This was my first thought as well, but I'm not sure exactly what would be causing this - asset hashes should remain consistent across systems. Is it possible your node version is somehow switching?

I recommend testing this on a self-hosted runner and seeing if the issue persists, that could potentially unblock you

**Reply by fynnfluegge** [reply]
Thanks for responding! Node/npm/yarn versions stays the same across deployments. I think I will use a self hosted runner as you recommend 👍 However, still curious what's the cause in github actions runner

**Reply by peterwoodworth** [reply]
Yeah I'm curious about that too, is there any chance you could setup a repo for this so I can try to reproduce it?

**Reply by fynnfluegge** [reply]
It happens in this repo 
👉https://github.com/fynnfluegge/rocketnotes

You can see in the "Deploy" actions that all lambda functions are updated every run. Doing cdk deploy on local machine does not update the lambda functions as expected.

**Reply by peterwoodworth** [reply]
Thanks @fynnfluegge, I'll see about reproducing this in my own repo at some point. Have you had any luck with, or been able to try the self hosted runner?

**Reply by fynnfluegge** [reply]
Thanks @peterwoodworth. I first tried to switch between Node version 17 and 18 but it has no effect. The aws-cdk global package is cached between runs by the way, so that version does not change. I don't see any more variables that could affect this for now, so going to switch to self hosted runner soon  ✅

**Reply by JannikWempe** [reply]
We recently realized the same. I could not recreate the changing hashes locally but they change in a GH Action causing very long running diffs.

@fynnfluegge I am curious if you have tried a self-hosted runner or found another solution.

**Reply by AlessandroVol23** [reply]
Hey @fynnfluegge did you had any success with a self-hosted runner? We're still facing this issue unfortunately.
Thanks!

**Reply by Exter-dg** [reply]
Our Lambdas get updates everytime from Codepipeline. Even when there is no change in the entire pipeline

**Reply by sushantmimani** [reply]
We're facing this issue too. We are on self-hosted runners but that doesn't seem to change the outcome. Found this but currently not able to re-work this for my use case.
https://github.com/hupe1980/jest-cdk-snapshot/issues/8

**Reply by fynnfluegge** [reply]
Maybe similar issue as it is in terraform with this as a workaround?
https://github.com/bboe/deterministic_zip

**Reply by sushantmimani** [reply]
My issue turned out to be that the `__pycache__` directory was in `.gitignore`. So it was missing on the runners, but present locally when calculating the hashes. I added this to my lambda creation and tests have been passing locally and on GHA.
```
  self.lambda_func = _alambda.PythonFunction(
            self,
            id_,
           ....
               asset_excludes=["__pycache__", "*.pyc", ".pytest_cache"],
            ),
          .....
```

To identify what was causing the issue, I ran `cdk deploy` from my local system, downloaded the contents from S3 and compared with the contents when `cdk deploy` runs from GHA.

**Comment by rehanvdm** [comment]
I explored it in depth for NodeJS here https://rehanvdm.com/blog/cdk-shorts-1-consistent-asset-hashing-nodejs

> TL;DR
>
> It is the wild-wild west within the node_modules directory, it mutates after installation and is the cause for non-deterministic hashing.
>
> The npm system, after installation, adds useless metadata fields (starting with underscore, specifically _where ) to each modules package.json, of which some are directory dependent. On top of that every module can add their own properties to the package.json. It is these directory dependent properties that are the root cause of non-deterministic hashes and deployments.
